### PR TITLE
storage/minio: chunk zip file

### PR DESCRIPTION
**Pull request by marwan-at-work**
Fixes https://github.com/gomods/athens/issues/1106

@manugupt1 let me know if this is okay for Minio. The more I look into it, the more I realize the Save implementation from Minio is correct and the module.Uploader is in fact incorrect. In other words, the other PR where we generalize the upload code should be reversed.

### CODE REVIEW

**Comment by manugupt1** [comment]
Hmm.. About the general one, it will be a breaking change, which is what I would like to avoid..

**Comment by marwan-at-work** [comment]
@manugupt1 the funny part is that we might have to break all storages except Minio :) but that's outside the scope of this PR

**Comment by manugupt1** [comment]
@marwan-at-work Which is exactly the other PR solves, I would like to understand why this one is better though.

**Comment by marpio** [comment]
@marwan-at-work
Are you referring to path under which the module bits are stored? Yeah, in order to be able to implement `Exists` by checking the count of elements equals 3 we would have to fix other storages or specifically the `PackageVersionedName` utility function. The minio implementation is actually the only one that uses the correct path - `module/version/{info,mod,zip}`
The question is do we want to introduce a breaking change and if so how to handle it.

**Comment by marpio** [comment]
Also related https://github.com/gomods/athens/issues/850

File: pkg/storage/minio/saver.go [hunk 1]
```diff
@@ -32,3 +37,74 @@ func (s *storageImpl) Save(ctx context.Context, module, vsn string, mod []byte,
...
+	zipFileName := dir + "/" + "source.zip"
+	dst, err := minio.NewDestinationInfo(s.bucketName, zipFileName, nil, nil)
+	if err != nil {
+		return errors.E(op, errors.E("minio.NewDestinationInfo", err))
+	}
+	err = s.minioClient.ComposeObject(dst, wr.srcs)
+	if err != nil {
+		return errors.E(op, errors.E("minio.ComposeObject", err))
+	}
+	err = s.removeParts(mod, ver, wr.numParts)
```
[the remark this answers is not in the document]
Comment (marwan-at-work) [reply pkg/storage/minio/saver.go]: @manugupt1 we cannot, because we have to remove each part individually. There's no concept of "remove directory"

File: pkg/storage/minio/saver.go [hunk 2]
```diff
@@ -32,3 +37,74 @@ func (s *storageImpl) Save(ctx context.Context, module, vsn string, mod []byte,
...
+	_, zw.err = zw.c.PutObject(zw.bucketName, partName, bytes.NewReader(p), plen, minio.PutObjectOptions{})
+	if zw.err != nil {
+		return 0, errors.E(op, zw.err)
+	}
+	zw.srcs = append(zw.srcs, minio.NewSourceInfo(zw.bucketName, partName, nil))
+	zw.numParts++
+	return len(p), nil
+}
+
+func (s *storageImpl) saveZip(dir, mod, ver string, zip io.Reader) error {
```
[the remark this answers is not in the document]
Comment (marwan-at-work) [reply pkg/storage/minio/saver.go]: @manugupt1 good point on locking, but our SingleFlight implementations already lock the entire save process so we should be safe.

**Comment by marwan-at-work** [comment]
@manugupt1 

> I would like to understand why this one is better though.

The other PR moves the code to module.Uploader, which is something we're trying to break anyway. Furthermore, to just be able to chunk the code for Minio, we had to change every single storage in our code base...that made not much sense to me. I'd rather keep weird Minio edge cases, to minio and not have to alter the whole codebase for it. 

>I would like to see some tests here for chunks but looks good to me

As long as the complaints test pass, we know we didn't break storage for anyone (hopefully our tests are thorough enough to guarantee that). 
That said, I'd like to add some unit tests to Minio where check that the "parts" are properly cleaned up...something the complaints tests shouldn't care about, but Athens should. 

### One last thing
It turns out we might need the Minio package at all: https://github.com/gomods/athens/issues/850

So I suggest we merge this to unblock people who are experiencing OOM kills and work on completely removing it if there are no differences between s3 and minio clients.

**Comment by marwan-at-work** [comment]
@marpio 
> The question is do we want to introduce a breaking change and if so how to handle it.

We can discuss that in the appropriate issue, I think we can probably write a script to move people's storage. We can also put forward compatibility hints in the storage itself. I have a few more ideas, but until then is this PR ok to unblock the issue that was created above?

**Comment by arschles** [comment]
@marwan-at-work is there an issue for this?

>@manugupt1 the funny part is that we might have to break all storages except Minio :) but that's outside the scope of this PR

If not, could you create one so we can talk about that there?

**Comment by marwan-at-work** [comment]
@arschles yes there is don't have it handy but can post it here later tongiht

**Comment by arschles** [comment]
ok, cool, thanks!
